kvserver: shard Raft scheduler

erikgrinaker opened the pull request:
The Raft scheduler mutex can become very contended on machines with many cores and high range counts. This patch shards the scheduler by allocating ranges and workers to individual scheduler shards.

By default, we create a new shard for every 16 workers, and distribute workers evenly. We spin up 8 workers per CPU core, capped at 96, so 16 is equivalent to 2 CPUs per shard, or a maximum of 6 shards. This significantly relieves contention at high core counts, while also avoiding starvation by excessive sharding. The shard size can be adjusted via `COCKROACH_SCHEDULER_SHARD_SIZE`.

This results in a substantial performance improvement on high-CPU nodes:

```
name                                    old ops/sec  new ops/sec  delta
kv0/enc=false/nodes=3/cpu=4              7.71k ± 5%   7.93k ± 4%     ~     (p=0.310 n=5+5)
kv0/enc=false/nodes=3/cpu=8              15.6k ± 3%   14.8k ± 7%     ~     (p=0.095 n=5+5)
kv0/enc=false/nodes=3/cpu=32             43.4k ± 2%   45.0k ± 3%   +3.73%  (p=0.032 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=aws   40.5k ± 2%   61.7k ± 1%  +52.53%  (p=0.008 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=gce   35.6k ± 4%   44.5k ± 0%  +24.99%  (p=0.008 n=5+5)

name                                    old p50      new p50      delta
kv0/enc=false/nodes=3/cpu=4               21.2 ± 6%    20.6 ± 3%     ~     (p=0.397 n=5+5)
kv0/enc=false/nodes=3/cpu=8               10.5 ± 0%    11.2 ± 8%     ~     (p=0.079 n=4+5)
kv0/enc=false/nodes=3/cpu=32              4.16 ± 1%    4.00 ± 5%     ~     (p=0.143 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=aws    3.00 ± 0%    2.00 ± 0%     ~     (p=0.079 n=4+5)
kv0/enc=false/nodes=3/cpu=96/cloud=gce    4.70 ± 0%    4.10 ± 0%  -12.77%  (p=0.000 n=5+4)

name                                    old p95      new p95      delta
kv0/enc=false/nodes=3/cpu=4               61.6 ± 5%    60.8 ± 3%     ~     (p=0.762 n=5+5)
kv0/enc=false/nodes=3/cpu=8               28.3 ± 4%    30.4 ± 0%   +7.34%  (p=0.016 n=5+4)
kv0/enc=false/nodes=3/cpu=32              7.98 ± 2%    7.60 ± 0%   -4.76%  (p=0.008 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=aws    12.3 ± 2%     6.8 ± 0%  -44.72%  (p=0.000 n=5+4)
kv0/enc=false/nodes=3/cpu=96/cloud=gce    10.2 ± 3%     8.9 ± 0%  -12.75%  (p=0.000 n=5+4)

name                                    old p99      new p99      delta
kv0/enc=false/nodes=3/cpu=4               89.8 ± 7%    88.9 ± 6%     ~     (p=0.921 n=5+5)
kv0/enc=false/nodes=3/cpu=8               46.1 ± 0%    48.6 ± 5%   +5.47%  (p=0.048 n=5+5)
kv0/enc=false/nodes=3/cpu=32              11.5 ± 0%    11.0 ± 0%   -4.35%  (p=0.008 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=aws    14.0 ± 3%    12.1 ± 0%  -13.32%  (p=0.008 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=gce    14.3 ± 3%    13.1 ± 0%   -8.55%  (p=0.000 n=4+5)

```

The basic cost of enqueueing ranges in the scheduler (without workers or contention) only increases slightly in absolute terms, thanks to `raftSchedulerBatch` pre-sharding the enqueued ranges:

```
name                                                                 old time/op  new time/op  delta
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=8-24         457ns ± 2%   564ns ± 2%   +23.36%  (p=0.001 n=7+7)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=16-24        461ns ± 3%   563ns ± 2%   +22.14%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=32-24        459ns ± 2%   591ns ± 2%   +28.63%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=64-24        455ns ± 0%   776ns ± 5%   +70.60%  (p=0.001 n=6+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=128-24       456ns ± 2%  1058ns ± 1%  +132.13%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=8-24    7.15ms ± 1%  8.18ms ± 1%   +14.33%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=16-24   7.13ms ± 1%  8.18ms ± 1%   +14.77%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=32-24   7.12ms ± 2%  7.86ms ± 1%   +10.30%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=64-24   7.20ms ± 1%  7.11ms ± 1%    -1.27%  (p=0.001 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=128-24  7.12ms ± 2%  7.16ms ± 3%      ~     (p=0.721 n=8+8)
```

Furthermore, on an idle 24-core 3-node cluster with 50.000 unquiesced ranges, this reduced CPU usage from 12% to 10%.

Resolves #98811.
Touches #96800.

Hat tip to @nvanbenschoten for the initial prototype.

Epic: none
Release note (performance improvement): The Raft scheduler is now sharded to relieve contention during range Raft processing, which can significantly improve performance at high CPU core counts.

erikgrinaker committed:
kvserver: shard Raft scheduler

The Raft scheduler mutex can become very contended on machines with many
cores and high range counts. This patch shards the scheduler by
allocating ranges and workers to individual scheduler shards.

By default, we create a new shard for every 16 workers, and distribute
workers evenly. We spin up 8 workers per CPU core, capped at 96, so 16
is equivalent to 2 CPUs per shard, or a maximum of 6 shards. This
significantly relieves contention at high core counts, while also
avoiding starvation by excessive sharding. The shard size can be
adjusted via `COCKROACH_SCHEDULER_SHARD_SIZE`.

This results in a substantial performance improvement on high-CPU nodes:

```
name                                    old ops/sec  new ops/sec  delta
kv0/enc=false/nodes=3/cpu=4              7.71k ± 5%   7.93k ± 4%     ~     (p=0.310 n=5+5)
kv0/enc=false/nodes=3/cpu=8              15.6k ± 3%   14.8k ± 7%     ~     (p=0.095 n=5+5)
kv0/enc=false/nodes=3/cpu=32             43.4k ± 2%   45.0k ± 3%   +3.73%  (p=0.032 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=aws   40.5k ± 2%   61.7k ± 1%  +52.53%  (p=0.008 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=gce   35.6k ± 4%   44.5k ± 0%  +24.99%  (p=0.008 n=5+5)

name                                    old p50      new p50      delta
kv0/enc=false/nodes=3/cpu=4               21.2 ± 6%    20.6 ± 3%     ~     (p=0.397 n=5+5)
kv0/enc=false/nodes=3/cpu=8               10.5 ± 0%    11.2 ± 8%     ~     (p=0.079 n=4+5)
kv0/enc=false/nodes=3/cpu=32              4.16 ± 1%    4.00 ± 5%     ~     (p=0.143 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=aws    3.00 ± 0%    2.00 ± 0%     ~     (p=0.079 n=4+5)
kv0/enc=false/nodes=3/cpu=96/cloud=gce    4.70 ± 0%    4.10 ± 0%  -12.77%  (p=0.000 n=5+4)

name                                    old p95      new p95      delta
kv0/enc=false/nodes=3/cpu=4               61.6 ± 5%    60.8 ± 3%     ~     (p=0.762 n=5+5)
kv0/enc=false/nodes=3/cpu=8               28.3 ± 4%    30.4 ± 0%   +7.34%  (p=0.016 n=5+4)
kv0/enc=false/nodes=3/cpu=32              7.98 ± 2%    7.60 ± 0%   -4.76%  (p=0.008 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=aws    12.3 ± 2%     6.8 ± 0%  -44.72%  (p=0.000 n=5+4)
kv0/enc=false/nodes=3/cpu=96/cloud=gce    10.2 ± 3%     8.9 ± 0%  -12.75%  (p=0.000 n=5+4)

name                                    old p99      new p99      delta
kv0/enc=false/nodes=3/cpu=4               89.8 ± 7%    88.9 ± 6%     ~     (p=0.921 n=5+5)
kv0/enc=false/nodes=3/cpu=8               46.1 ± 0%    48.6 ± 5%   +5.47%  (p=0.048 n=5+5)
kv0/enc=false/nodes=3/cpu=32              11.5 ± 0%    11.0 ± 0%   -4.35%  (p=0.008 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=aws    14.0 ± 3%    12.1 ± 0%  -13.32%  (p=0.008 n=5+5)
kv0/enc=false/nodes=3/cpu=96/cloud=gce    14.3 ± 3%    13.1 ± 0%   -8.55%  (p=0.000 n=4+5)
```

Furthermore, on an idle 24-core 3-node cluster with 50.000 unquiesced
ranges, this reduced CPU usage from 12% to 10%.

The basic cost of enqueueing ranges in the scheduler (without workers or
contention) only increases slightly in absolute terms, thanks to
`raftSchedulerBatch` pre-sharding the enqueued ranges:

```
name                                                                 old time/op  new time/op  delta
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=8-24         457ns ± 2%   564ns ± 2%   +23.36%  (p=0.001 n=7+7)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=16-24        461ns ± 3%   563ns ± 2%   +22.14%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=32-24        459ns ± 2%   591ns ± 2%   +28.63%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=64-24        455ns ± 0%   776ns ± 5%   +70.60%  (p=0.001 n=6+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=128-24       456ns ± 2%  1058ns ± 1%  +132.13%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=8-24    7.15ms ± 1%  8.18ms ± 1%   +14.33%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=16-24   7.13ms ± 1%  8.18ms ± 1%   +14.77%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=32-24   7.12ms ± 2%  7.86ms ± 1%   +10.30%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=64-24   7.20ms ± 1%  7.11ms ± 1%    -1.27%  (p=0.001 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=128-24  7.12ms ± 2%  7.16ms ± 3%      ~     (p=0.721 n=8+8)
```

Epic: none
Release note (performance improvement): The Raft scheduler is now
sharded to relieve contention during range Raft processing, which can
significantly improve performance at high CPU core counts.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98854)
<!-- Reviewable:end -->

erikgrinaker commented:
Going to run `kv0/enc=false/nodes=3/cpu=96`. Meanwhile, here's some mutex profiles from 50k unquiesced ranges on a 48-core machine:

```
$ go tool pprof -diff_base mutex-before.pb.gz mutex-after.pb.gz
(pprof) top -cum
Showing nodes accounting for -214.21s, 54.53% of 392.83s total
Dropped 269 nodes (cum <= 1.96s)
Showing top 10 nodes out of 23
      flat  flat%   sum%        cum   cum%
         0     0%     0%   -215.05s 54.74%  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
  -214.21s 54.53% 54.53%   -214.21s 54.53%  sync.(*Mutex).Unlock
         0     0% 54.53%   -212.96s 54.21%  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker
         0     0% 54.53%    173.44s 44.15%  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2
         0     0% 54.53%     39.20s  9.98%  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick
         0     0% 54.53%     38.29s  9.75%  github.com/cockroachdb/cockroach/pkg/util/hlc.(*Clock).NowAsClockTimestamp
         0     0% 54.53%     38.29s  9.75%  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).tick
         0     0% 54.53%    -18.72s  4.77%  sync.(*Cond).Wait
```

Before:

<img width="1446" alt="Screenshot 2023-03-17 at 14 23 27" src="https://user-images.githubusercontent.com/644420/225917206-6b54e988-a8be-4119-882f-671d2fc66524.png">

After:

<img width="1446" alt="Screenshot 2023-03-17 at 14 23 40" src="https://user-images.githubusercontent.com/644420/225917270-e342c7a3-b881-4fc5-822b-0f5d49b13f4a.png">

erikgrinaker commented:
The results are pretty good (see PR description), as expected over in #96800.

Given the upside, I'm inclined to merge this before the branch cut. The changes are fairly well-contained, so the primary risk would be worker starvation with uneven shard load, but with 12 workers per shard the risk seems fairly minor. Let's have a look at the nightly benchmarks.

erikgrinaker commented:
Note to self: consider adjusting the tick loop and heartbeat uncoalescing to pre-shard the inputs and enqueue them directly to the shard, avoiding the O(shards) looping.

erikgrinaker commented:
I added a commit with `raftSchedulerBatch` which avoids the O(shards) factor in `enqueueN` scans. However, on a 24-core cluster with 50k unquiesced ranges it didn't have any measurable impact on CPU usage, as one might expect. Unclear whether it's really an improvement, wdyt?

erikgrinaker committed:
kvserver: add `BenchmarkSchedulerEnqueueRaftTicks`

Epic: none
Release note: None

erikgrinaker commented:
Added a microbenchmark for `EnqueueRaftTicks()`, varying the range count and worker count (which determines the shard count by a factor of 1/12).  The results below are for the `collect=true` case where we include the cost of iterating over the replica map to collect range IDs. This showed that the pre-sharding via `raftSchedulerBatch` saved enough to be worthwhile.

Sharded scheduler without `raftSchedulerBatch`, compared to the unsharded scheduler:

```
name                                                                 old time/op  new time/op   delta
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=8-24         457ns ± 2%    565ns ± 3%   +23.65%  (p=0.000 n=7+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=16-24        461ns ± 3%    561ns ± 2%   +21.71%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=32-24        459ns ± 2%    679ns ± 3%   +47.91%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=64-24        455ns ± 0%   1086ns ± 2%  +138.97%  (p=0.001 n=6+7)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=128-24       456ns ± 2%   1838ns ± 1%  +303.51%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=8-24    7.15ms ± 1%   8.09ms ± 1%   +13.05%  (p=0.000 n=8+7)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=16-24   7.13ms ± 1%   8.06ms ± 1%   +13.06%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=32-24   7.12ms ± 2%  10.07ms ± 0%   +41.34%  (p=0.001 n=8+6)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=64-24   7.20ms ± 1%  12.59ms ± 1%   +74.87%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=128-24  7.12ms ± 2%  17.10ms ± 1%  +140.14%  (p=0.000 n=8+8)
```

Pre-sharding via `raftSchedulerBatch`, compared to the unsharded scheduler:

```
name                                                                 old time/op  new time/op  delta
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=8-24         457ns ± 2%   564ns ± 2%   +23.36%  (p=0.001 n=7+7)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=16-24        461ns ± 3%   563ns ± 2%   +22.14%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=32-24        459ns ± 2%   591ns ± 2%   +28.63%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=64-24        455ns ± 0%   776ns ± 5%   +70.60%  (p=0.001 n=6+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=10/workers=128-24       456ns ± 2%  1058ns ± 1%  +132.13%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=8-24    7.15ms ± 1%  8.18ms ± 1%   +14.33%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=16-24   7.13ms ± 1%  8.18ms ± 1%   +14.77%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=32-24   7.12ms ± 2%  7.86ms ± 1%   +10.30%  (p=0.000 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=64-24   7.20ms ± 1%  7.11ms ± 1%    -1.27%  (p=0.001 n=8+8)
SchedulerEnqueueRaftTicks/collect=true/ranges=100000/workers=128-24  7.12ms ± 2%  7.16ms ± 3%      ~     (p=0.721 n=8+8)
```

The performance improvement with higher worker counts at `ranges=100000` is interesting though. I haven't tracked it down, but my guess is that it's related to the size of the per-shard `state` map (i.e. hash collisions).

I've squashed the `raftSchedulerBatch` commit into the main commit here.

pavelkalinnikov commented on pkg/kv/kvserver/scheduler_test.go:
```diff
@@ -249,9 +249,14 @@ func TestSchedulerLoop(t *testing.T) {
+
+	batch := s.NewEnqueueBatch()
+	defer batch.Close()
+	batch.Add(1)
+	batch.Add(2)
+	batch.Add(3)
```
How about a variadic `Add`? OTOH, fine as is, since the prod code `Add`s one by one, so it's better to avoid hidden slices in it.

erikgrinaker replied:
Doesn't seem worth it.

pav-kv reviewed (approved):
LGTM. All nits are optional.

erikgrinaker commented:
Looks like this is regressing slightly at lower core counts. Going to do another set with shard size 16 instead of 12.

```
name                          old ops/sec  new ops/sec  delta
kv0/enc=false/nodes=3/cpu=4    7.71k ± 5%   7.98k ± 2%     ~     (p=0.095 n=5+5)
kv0/enc=false/nodes=3/cpu=8    15.6k ± 3%   15.5k ± 1%     ~     (p=0.310 n=5+5)
kv0/enc=false/nodes=3/cpu=32   43.4k ± 2%   42.7k ± 5%     ~     (p=0.310 n=5+5)

name                          old p50      new p50      delta
kv0/enc=false/nodes=3/cpu=4     21.2 ± 6%    20.3 ± 3%     ~     (p=0.238 n=5+5)
kv0/enc=false/nodes=3/cpu=8     10.5 ± 0%    10.8 ± 3%     ~     (p=0.238 n=4+5)
kv0/enc=false/nodes=3/cpu=32    4.16 ± 1%    4.24 ± 8%     ~     (p=0.683 n=5+5)

name                          old p95      new p95      delta
kv0/enc=false/nodes=3/cpu=4     61.6 ± 5%    61.6 ± 2%     ~     (p=1.000 n=5+5)
kv0/enc=false/nodes=3/cpu=8     28.3 ± 4%    28.7 ± 2%     ~     (p=0.643 n=5+5)
kv0/enc=false/nodes=3/cpu=32    7.98 ± 2%    8.02 ± 1%     ~     (p=1.000 n=5+5)

name                          old p99      new p99      delta
kv0/enc=false/nodes=3/cpu=4     89.8 ± 7%    88.1 ± 0%     ~     (p=0.167 n=5+5)
kv0/enc=false/nodes=3/cpu=8     46.1 ± 0%    47.4 ± 3%     ~     (p=0.167 n=5+5)
kv0/enc=false/nodes=3/cpu=32    11.5 ± 0%    12.8 ± 6%  +11.30%  (p=0.008 n=5+5)
```

pavelkalinnikov commented:
@erikgrinaker 
> Looks like this is regressing slightly at lower core counts. Going to do another set with shard size 16 instead of 12.

When you converge on some default shard/workers sizes, please add them to the unit test.

nvb reviewed (commented):
Nice!

> Reviewed 3 of 3 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-:-NR-3M836wS9a7QiUyY7:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @pavelkalinnikov, and @tbg)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NQzsbL5DLAYBVP1QxNX):*
nit: could you include the benchmark results in the commit message?
___
*[`pkg/kv/kvserver/scheduler.go` line 329 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-1sEVDcUHqWNQlqWs:-NR-1sEVDcUHqWNQlqWt:b-jzyufq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5545ec4f00125a5dc98011ed64d87d54a0d2aff8/pkg/kv/kvserver/scheduler.go#L329)):*
> ```Go
> }
> 
> func (s *raftScheduler) worker(ctx context.Context, shard *raftSchedulerShard) {
> ```

Did you consider moving this to be a method on `*raftSchedulerShard`? Does the method need access to the full scheduler?
___
*[`pkg/kv/kvserver/scheduler.go` line 471 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-2UJdB8L3zANxPhAI:-NR-2UJdB8L3zANxPhAJ:b-xdqveh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5545ec4f00125a5dc98011ed64d87d54a0d2aff8/pkg/kv/kvserver/scheduler.go#L471)):*
> ```Go
> }
> 
> func (s *raftSchedulerShard) enqueueN(addFlags raftScheduleFlags, ids ...roachpb.RangeID) int {
> ```

nit: do we want to use a different name for the `raftSchedulerShard` receivers? Maybe `ss`? While reading this file, it's easy to forget which scope a method is operating on.
___
*[`pkg/kv/kvserver/scheduler_test.go` line 258 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NQjhOfbxYaxWO8l0HtR-r2-258:-NR-36ii2oTOVPZsDXVm:by5zfum) ([raw file](https://github.com/cockroachdb/cockroach/blob/500dc1c23d3cdffca00759803610e9671a70bc11/pkg/kv/kvserver/scheduler_test.go#L258)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Doesn't seem worth it.
</blockquote></details>

+1 to not being worth it, especially if it comes with any runtime cost.
___
*[`pkg/kv/kvserver/scheduler_test.go` line 397 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NQzsuFoD89A3DPXTwW7:-NQzsuFoD89A3DPXTwW8:b-2nvbkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafdb52f2e5e7a8234dcf64c50b2f298169c50af/pkg/kv/kvserver/scheduler_test.go#L397)):*
> ```Go
> 		}
> 		s.EnqueueRaftTicks(ids...)
> 		s.mu.queue = rangeIDQueue{} // flush queue
> ```

nit: consider mentioning that we haven't started the raftScheduler, so this won't race with workers pulling off the queue.
___
*[`pkg/kv/kvserver/scheduler_test.go` line 397 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NQztKKj-NSN3AKA6WrP:-NQztKKkDIuJZAJXHo9O:b-y2ovdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafdb52f2e5e7a8234dcf64c50b2f298169c50af/pkg/kv/kvserver/scheduler_test.go#L397)):*
> ```Go
> 		}
> 		s.EnqueueRaftTicks(ids...)
> 		s.mu.queue = rangeIDQueue{} // flush queue
> ```

^ raises the question about whether this is measuring the true cost of mutex contention in the scheduler, because we're not seeing contended mutexes and we're not seeing cache misses. I don't see a good way to measure that additional cost and still have a stable benchmark.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
I did another run with 16 workers per shard, and the results are slightly better at lower core counts, but it's hard to say really with the variation here. The below results are compared to `master`. I've bumped the default to 16 to be conservative. I'll rerun the 96-core results as well.

```
name                          old ops/sec  new ops/sec  delta
kv0/enc=false/nodes=3/cpu=4    7.71k ± 5%   7.93k ± 4%    ~     (p=0.310 n=5+5)
kv0/enc=false/nodes=3/cpu=8    15.6k ± 3%   14.8k ± 7%    ~     (p=0.095 n=5+5)
kv0/enc=false/nodes=3/cpu=32   43.4k ± 2%   45.0k ± 3%  +3.73%  (p=0.032 n=5+5)

name                          old p50      new p50      delta
kv0/enc=false/nodes=3/cpu=4     21.2 ± 6%    20.6 ± 3%    ~     (p=0.397 n=5+5)
kv0/enc=false/nodes=3/cpu=8     10.5 ± 0%    11.2 ± 8%    ~     (p=0.079 n=4+5)
kv0/enc=false/nodes=3/cpu=32    4.16 ± 1%    4.00 ± 5%    ~     (p=0.143 n=5+5)

name                          old p95      new p95      delta
kv0/enc=false/nodes=3/cpu=4     61.6 ± 5%    60.8 ± 3%    ~     (p=0.762 n=5+5)
kv0/enc=false/nodes=3/cpu=8     28.3 ± 4%    30.4 ± 0%  +7.34%  (p=0.016 n=5+4)
kv0/enc=false/nodes=3/cpu=32    7.98 ± 2%    7.60 ± 0%  -4.76%  (p=0.008 n=5+5)

name                          old p99      new p99      delta
kv0/enc=false/nodes=3/cpu=4     89.8 ± 7%    88.9 ± 6%    ~     (p=0.921 n=5+5)
kv0/enc=false/nodes=3/cpu=8     46.1 ± 0%    48.6 ± 5%  +5.47%  (p=0.048 n=5+5)
kv0/enc=false/nodes=3/cpu=32    11.5 ± 0%    11.0 ± 0%  -4.35%  (p=0.008 n=5+5)
```

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98854)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @pavelkalinnikov, and @tbg)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NQzsbL5DLAYBVP1QxNX):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: could you include the benchmark results in the commit message?
</blockquote></details>

Yep, done.
___
*[`pkg/kv/kvserver/scheduler.go` line 329 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-1sEVDcUHqWNQlqWs:-NR-5_Pl4hvBkgl26eW7:bex0qkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5545ec4f00125a5dc98011ed64d87d54a0d2aff8/pkg/kv/kvserver/scheduler.go#L329)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you consider moving this to be a method on `*raftSchedulerShard`? Does the method need access to the full scheduler?
</blockquote></details>

Yes, I considered it, but I wanted to keep the diff minimal to convince ourselves that it was reasonably correct. There doesn't seem to be much to gain from it, so I'll leave it like this for now, but worth reconsidering.
___
*[`pkg/kv/kvserver/scheduler.go` line 471 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-2UJdB8L3zANxPhAI:-NR-5omj-qksLe83N9S_:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5545ec4f00125a5dc98011ed64d87d54a0d2aff8/pkg/kv/kvserver/scheduler.go#L471)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: do we want to use a different name for the `raftSchedulerShard` receivers? Maybe `ss`? While reading this file, it's easy to forget which scope a method is operating on.
</blockquote></details>

Good idea, done.
___
*[`pkg/kv/kvserver/scheduler_test.go` line 397 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NQzsuFoD89A3DPXTwW7:-NR-5xYHBCiLmw0POxrz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafdb52f2e5e7a8234dcf64c50b2f298169c50af/pkg/kv/kvserver/scheduler_test.go#L397)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: consider mentioning that we haven't started the raftScheduler, so this won't race with workers pulling off the queue.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/scheduler_test.go` line 397 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NQztKKj-NSN3AKA6WrP:-NR-5yMT7sgX4wI2O4Ue:b-ghq6e7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafdb52f2e5e7a8234dcf64c50b2f298169c50af/pkg/kv/kvserver/scheduler_test.go#L397)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

^ raises the question about whether this is measuring the true cost of mutex contention in the scheduler, because we're not seeing contended mutexes and we're not seeing cache misses. I don't see a good way to measure that additional cost and still have a stable benchmark.
</blockquote></details>

Yeah, it isn't intended to measure mutex contention, only to measure the basis cost of enqueueing ranges in the scheduler. It was motivated by finding out whether `raftSchedulerBatch` was even worth it, or if a naïve filtering loop would do.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-:-NR-ickt4wbI61XeO8lG:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @pavelkalinnikov, and @tbg)

___
*[`-- commits` line 15 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-jGMY9dk18W4Thbc0):*
I wonder if we need this cap anymore. It was added back in 277cc156fd04ed4ba93bf336023727f5fe7f916b to fight collapse under mutation contention. On the other hand, we no longer block on fsync under the raft scheduler, so there's an argument that the worker count is too high now. Not worth changing in either direction right now.
___
*[`-- commits` line 35 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-iye68f5l5VZCwe7V):*
> 3.00 ± 0%    2.00 ± 0%     ~     (p=0.079 n=4+5)

We're just missing out on a statistically significant major improvement. That's too bad.
___
*[`pkg/kv/kvserver/scheduler.go` line 329 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-1sEVDcUHqWNQlqWs:-NR-iE1nBpi4a-K2O5Qo:b-4h6su9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5545ec4f00125a5dc98011ed64d87d54a0d2aff8/pkg/kv/kvserver/scheduler.go#L329)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Yes, I considered it, but I wanted to keep the diff minimal to convince ourselves that it was reasonably correct. There doesn't seem to be much to gain from it, so I'll leave it like this for now, but worth reconsidering.
</blockquote></details>

Mind leaving a TODO to clean this up then? And also `s/shard/ss/g` in this function and elsewhere for consistent naming.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98854)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten, @pavelkalinnikov, and @tbg)

___
*[`-- commits` line 15 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-jGMY9dk18W4Thbc0):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I wonder if we need this cap anymore. It was added back in 277cc156fd04ed4ba93bf336023727f5fe7f916b to fight collapse under mutation contention. On the other hand, we no longer block on fsync under the raft scheduler, so there's an argument that the worker count is too high now. Not worth changing in either direction right now.
</blockquote></details>

Yeah, I've been wondering the same thing, but don't want to start fiddling with it now. Wrote up #99063.
___
*[`-- commits` line 35 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-iye68f5l5VZCwe7V):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> 3.00 ± 0%    2.00 ± 0%     ~     (p=0.079 n=4+5)

We're just missing out on a statistically significant major improvement. That's too bad.
</blockquote></details>

I know! I was tempted to rerun these with higher counts to get those juicy precentages, but alas, we'll have to settle for a statistically insignificant 33% improvement.
___
*[`pkg/kv/kvserver/scheduler.go` line 329 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98854#-NR-1sEVDcUHqWNQlqWs:-NR-o2kP-gd-ATWcBMYK:b-zar5wz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5545ec4f00125a5dc98011ed64d87d54a0d2aff8/pkg/kv/kvserver/scheduler.go#L329)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Mind leaving a TODO to clean this up then? And also `s/shard/ss/g` in this function and elsewhere for consistent naming.
</blockquote></details>

Did this now, on second thought. It avoids accidental misuse of other shards, and there were only a couple of dependencies to inject. Left the code in the same spot though, for history and ease of review.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
bors r+
